Update to webrtc v56

getkey opened the pull request:
The build process now [relies on the tool `gn`](https://webrtc.org/native-code/development/) whereas version 50 [used `gyp`](https://web.archive.org/web/20160323234502/http://webrtc.org/native-code/development/).

getkey committed:
Update to webrtc v56

The build process now relies on the tool `gn` whereas version 50 used `gyp`.

getkey commented:
This is not ready to be merged as node-webrtc fails to link against it:

```
  SOLINK_MODULE(target) Release/obj.target/wrtc.node
../third_party/webrtc/lib/libwebrtc.a(circular_buffer.o): In function `webrtc::CircularBuffer::~CircularBuffer()':
../../webrtc/modules/audio_processing/echo_detector/circular_buffer.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): multiple definition of `webrtc::CircularBuffer::~CircularBuffer()'
../third_party/webrtc/lib/libwebrtc.a(utility.o):../../webrtc/modules/audio_coding/test/utility.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): first defined here
../third_party/webrtc/lib/libwebrtc.a(circular_buffer.o): In function `webrtc::CircularBuffer::~CircularBuffer()':
../../webrtc/modules/audio_processing/echo_detector/circular_buffer.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): multiple definition of `webrtc::CircularBuffer::~CircularBuffer()'
../third_party/webrtc/lib/libwebrtc.a(utility.o):../../webrtc/modules/audio_coding/test/utility.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): first defined here
collect2: error: ld returned 1 exit status
make: *** [wrtc.target.mk:189: Release/obj.target/wrtc.node] Error 1
```

We don't need the definition in `webrtc/modules/audio_coding/test/utility.h` since it is used only by tests but I'm not really sure how to prevent it from being merged to `libwebrtc.a`.

markandrus commented:
Thanks, I hope to take a close look at this later this week. From what I
remember, those errors may be occurring due to how the object files are
being linked.

Also, any chance you've tested this out on Linux, Windows, and OS X? I
wanted to update this before, but saw some issues compiling on Windows (I
think there were bugs introduced to the Windows build steps, since
internally Google uses a slightly different set of steps). They may be
resolved now.

On Mon, Dec 19, 2016 at 6:32 AM Julien Mourer <notifications@github.com>
wrote:

> This is not ready to be merged as node-webrtc fails to link against it:
>
>
>   SOLINK_MODULE(target) Release/obj.target/wrtc.node
>
> ../third_party/webrtc/lib/libwebrtc.a(circular_buffer.o): In function `webrtc::CircularBuffer::~CircularBuffer()':
>
> ../../webrtc/modules/audio_processing/echo_detector/circular_buffer.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): multiple definition of `webrtc::CircularBuffer::~CircularBuffer()'
>
> ../third_party/webrtc/lib/libwebrtc.a(utility.o):../../webrtc/modules/audio_coding/test/utility.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): first defined here
>
> ../third_party/webrtc/lib/libwebrtc.a(circular_buffer.o): In function `webrtc::CircularBuffer::~CircularBuffer()':
>
> ../../webrtc/modules/audio_processing/echo_detector/circular_buffer.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): multiple definition of `webrtc::CircularBuffer::~CircularBuffer()'
>
> ../third_party/webrtc/lib/libwebrtc.a(utility.o):../../webrtc/modules/audio_coding/test/utility.cc:(.text._ZN6webrtc14CircularBufferD2Ev+0x0): first defined here
>
> collect2: error: ld returned 1 exit status
>
> make: *** [wrtc.target.mk:189: Release/obj.target/wrtc.node] Error 1
>
>
>
> We don't need the definition in webrtc/modules/audio_coding/test/utility.h
> since it is used only by tests but I'm not really sure how to prevent it
> from being merged to libwebrtc.a.
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/markandrus/build-webrtc/pull/2#issuecomment-267855444>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAV68myCFeHdHk1vtLcUsn7_-gh24ncBks5rJcKTgaJpZM4LQQmS>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

getkey committed:
Do not build tests

If they are built they get merged into `libwebrtc.a` which make it uselessly larger and may cause linking issues if a test re-exports a symbol.

getkey commented:
The `gn` config file defines a switch to disable building tests so this is fixed now, `node-webrtc` builds fine (after replacing [`rtc::Buffer` with `rtc::CopyOnWriteBuffer`](https://github.com/js-platform/node-webrtc/issues/262#issuecomment-238083733)). :-)

However running the ping-pong example fails:
```
~/d/node-webrtc> node examples/ping-pong-test.js
module.js:598
  return process.dlopen(module, path._makeLong(filename));
                 ^

Error: /home/getkey/devel/node-webrtc/build/wrtc/v0.0.62/Release/node-v51-linux-x64/wrtc.node: undefined symbol: _ZN3rtc6Thread5ClearEPNS_14MessageHandlerEjPSt4listINS_7MessageESaIS4_EE
    at Object.Module._extensions..node (module.js:598:18)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/getkey/devel/node-webrtc/lib/peerconnection.js:6:15)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
```

I've only built on Linux x64, but I'll try on Windows and report back. Unfortunately I don't have access to an OS X machine.

getkey commented on jakelib/gn.js:
```diff
@@ -0,0 +1,24 @@
+    stdio: 'inherit'
+  }, execOptions);
+
+  var cmd = GN + ' gen ' + cwd + ' --args=\'is_debug=' + (config.CONFIGURATION !== 'Release') + ' rtc_include_tests=false\'';
```
On windows this fails because single quotes are not supported in the command line.
I can either replace them with double quotes or use `execFileSync`, which seems more robust. In that case I think we could use `execFileSync` everywhere. What do you think?

Unfortunately that won't be enough to fix the build on Windows, it fails later with the following error:
```
Running ninja

ninja: error: obj/build/config/sanitizers/options_sources.ninja:9: expected newl
ine, got ':'

ninja: Entering directory `\build-webrtc\webrtc\src\out\Release_x64'
jake aborted.
Error: Command failed: \build-webrtc\depot_tools\ninja -C \build-webrtc\webrtc\s
rc\out\Release_x64
    at checkExecSyncError (child_process.js:490:13)
    at execSync (child_process.js:530:13)
(See full trace by running task with --trace)
```

getkey committed:
Use `execFileSync` instead of `execSync`

This fixes errors on Windows.

wildekek commented:
You're doing god's work @getkey. Is there a reason this is not merged yet? Not being able to update WebRTC is running into an innovation problem with server side and embedded webrtc development. I see more node projects(e.g. https://github.com/HefnySco/node-webrtc-native) using build-webrtc and moving the development forward would be such a good thing for all of us.

getkey commented:
It's not merged because [node-webrtc](https://github.com/js-platform/node-webrtc) panics with v56. I fixed a few things in https://github.com/js-platform/node-webrtc/pull/300 but there is one thing that I was unfortunately unable to fix. I'm pretty sure an experienced Node Native developer would find a solution though!

markandrus commented:
I want to see about using the great work in `libwebrtc` instead of this project. It's pretty painful supporting this build. I'm currently trying to land Node 9 and M57 support [here](https://github.com/js-platform/node-webrtc/pull/333).

getkey commented:
Nice, good to hear!
